lib/semantics/check-select-stmt.cc

```diff
@@ -0,0 +1,103 @@
+//= check-select-stmt.cc - Checker for select-case, select-rank, select-type ==
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//=--------------------------------------------------------------------------==
+
+#include "check-select-stmt.h"
+#include "tools.h"
+#include "../parser/message.h"
+
+namespace Fortran::semantics {
+
+void SelectStmtChecker::Leave(
+    const parser::CaseConstruct &selectcaseConstruct) {
+  const auto &selectcaseStmt{
+      std::get<parser::Statement<parser::SelectCaseStmt>>(
+          selectcaseConstruct.t)};
+  const auto &parsedExpr{
+      std::get<parser::Scalar<parser::Expr>>(selectcaseStmt.statement.t).thing};
+  TypeCategory selectcasestmttype;
+  if (const auto *expr{GetExpr(parsedExpr)}) {
+    if (auto type{expr->GetType()}) {
+      selectcasestmttype = type->category();
+      if ((selectcasestmttype != TypeCategory::Integer) &&
+          (selectcasestmttype != TypeCategory::Character) &&
+          (selectcasestmttype != TypeCategory::Logical)) {
+        // C1145 case-expr shall be of type character, integer, or logical.
+        context_.Say(parsedExpr.source,
+            "Select case expression must be of type character, integer, or logical"_err_en_US);
+      }
+    }
+  }
```

Comment:
What happens if selectcasestmttype has not been set in the if statement?

Reply:
I had the same question. But I didn't think of a scenario where the select statement is not having any type. It will be either intrinsic type or derived type. In both cases the variable will be set to some type. If you are aware of any scenario, please do let me know. I will check and do the necessary changes.

Reply:
If there is an error in the expression, GetExpr or GetType might not return a value.
Also, if the expression is a BOZLiteral it doesn't have a type. I'm not sure what should happen in that case -- maybe it should be an error.

Reply:
There are other expressions that don't have types, like the name of a subroutine.  So you need to ensure that `selectCaseStmtType` is not left uninitialized.  You could make `selectCaseStmtType` a `std::optional<>` to make it clear when it is set and when it is not.

Reply:
Thanks for the suggestion Klausler, Now I have made selectCaseStmtType a std::optional<>.